File: code/python/剑指offer26.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSubStructure(self, A: TreeNode, B: TreeNode) -> bool:
        if B is None:
            return False

        def isSub(a, b):
            if (a is None and b is not None) or (a is not None and b is None):
                return False
            if a.val != b.val:
                return False
            left = True
            if b.left is not None:
                left = isSub(a.left, b.left)
            right = True
            if b.right is not None:
                right = isSub(a.right, b.right)
            return left & right

        def fun(A, B):
            if A is None:
                return False
            if isSub(A, B) == True:
                return True
            else:
                return fun(A.left, B) | fun(A.right, B)

        return fun(A, B)

File: code/java/剑指offer26.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSub(TreeNode a,TreeNode b){
        if((a==null&&b!=null)||(a!=null&&b==null)) return false;
        if(a.val!=b.val) return false;
        boolean leftRes = true;
        if(b.left!=null) leftRes = isSub(a.left,b.left);
        boolean rightRes = true;
        if(b.right!=null) rightRes = isSub(a.right,b.right);
        return leftRes&&rightRes;
    }

    public boolean judge(TreeNode a,TreeNode b){
        if(a==null) return false;
        if(isSub(a,b)) return true;
        else return judge(a.left,b)|judge(a.right,b);
    }

    public boolean isSubStructure(TreeNode A, TreeNode B) {
        if(B==null) return false;
        return judge(A,B);
    }
}